### DISCUSSION TITLE
/hyndsight/tsoutliers/

### DISCUSSION THREAD

**Discussion by robjhyndman** [robjhyndman/robjhyndman.com#325] [Announcements]
Detecting time series outliers 
 https://robjhyndman.com/hyndsight/tsoutliers/

**Comment by robjhyndman** [comment]
```header
Originally posted by Andrew on 2021-09-06 13:19:36
````
Many thanks for this. Will these functions be implemented in the fpp3
set of packages? Thanks.

**Reply by robjhyndman** [reply]
```header
Originally posted on 2021-09-06 23:09:23
````
We are planning to add some outlier detection methods to fable, for
different types of models (not just STL).

**Reply by robjhyndman** [reply]
```header
Originally posted by Andrew on 2021-09-07 07:44:09
````
Thanks for the reply Rob. That\'s great to hear.

**Reply by robjhyndman** [reply]
```header
Originally posted by August on 2021-11-22 09:47:56
````
Hi Rob, out of interest would one of these additional models be
isolation forests? I am finding it (in my job) to often result in better
RMSE of test sets compared to profiling methods. That is, of course,
when it is paired with features such as datetime, holidays and lags etc.
It would be great to have in fable!

**Reply by robjhyndman** [reply]
```header
Originally posted on 2021-11-22 11:04:59
````
No, sorry, that\'s not on the agenda. But we are planning on having
tools implemented that will work with any fable model, and fable is
designed to be extended. So if someone wrote a forest extension to
fable, it should be easy to then use the (not yet implemented) outlier
detection methods.

**Reply by robjhyndman** [reply]
```header
Originally posted by BHEEMANNA on 2021-11-22 11:08:18
````
Hello sir, I have population data that is nonstationary then we check
stationary from the ADF test with get stationary by third difference but
we used Arima function in r, we get the stationary by second difference.
which one is the best model sir, please suggest to me, sir.\
Bheemanna\
Research Scholar\
Dept. of Statistics\
Karnatak University\'s\
Karnatak Arts College,\
Dharwad-580001.\
Karnataka.\
INDIA.

**Reply by robjhyndman** [reply]
```header
Originally posted on 2021-11-22 11:17:25
````
This is not a help site. Please ask such questions at
[crossvalidated.com](http://crossvalidated.com &quot;crossvalidated.com&quot;).

**Reply by robjhyndman** [reply]
```header
Originally posted by August on 2021-11-23 09:44:46
````
Really appreciate the response on this Rob, thank you. I may attempt a
forest extension, it will be a good learning experience if I can get the
time.

**Comment by robjhyndman** [comment]
```header
Originally posted by Phil Karl on 2021-09-10 08:19:01
````
[https://uploads.disquscdn.c\...](https://uploads.disquscdn.com/images/b21ac1d6bd90c87f71a03f6c8f854eabd44013b0b8e573691daf28be83fab367.png)
Thanks a lot! Is the seasonality identified automatically?\
In one of my cases it seems to cut away too much, can I adjust this?

**Reply by robjhyndman** [reply]
```header
Originally posted by Phil Karl on 2021-09-10 08:25:25
````
I found the solution by myself. Setting frequency = 52 in ts(\...)
helped. Unfortunately the penultimate peak is not identified as outlier
then.
[https://uploads.disquscdn.c\...](https://uploads.disquscdn.com/images/b968fa0bff4061264119d0950844b4ac9a5e39c19787c0192fec4d7c63ffe431.png)

**Reply by robjhyndman** [reply]
```header
Originally posted by Phil Karl on 2021-09-10 08:32:25
````
Setting frequency=52 in ts(\...) and lambda = &quot;auto&quot; in tsclean
helped:
[https://uploads.disquscdn.c\...](https://uploads.disquscdn.com/images/d9c47995ca728fba1bd009beb649a5ea684e589bdfa9a68fc95f80783cf42421.png)

**Comment by robjhyndman** [comment]
```header
Originally posted by BHEEMANNA on 2021-10-11 09:03:58
````
Hi sir, I am Bheemanna, and how can estimate the Arima polynomial
coefficients.

**Comment by robjhyndman** [comment]
```header
Originally posted by BHEEMANNA on 2021-10-11 09:05:01
````
Please help me sir

**Comment by wurli** [comment]
Hi Prof. Hyndman,

It doesn't seem like this has yet been implemented in the tidyverts - am I correct? Something seems to have gone awry migrating the above comments to Giscus, so it's difficult to tell how up to date your comments are - and which ones are from you!

Thanks,

Jacob

**Reply by robjhyndman** [reply]
Yes, I think the migration has screwed up somewhere. Too bad. Anyway, we are planning a more general approach in the tidyverts packages, where outlier detection is an additional function on top of any decomposition method. But other features are taking priority right now. It is easy enough to do though -- just get the components from an STL model, compute the IQR of the remainder, and identify where the remainder is larger than 3IQR in absolute value.

**Reply by wurli** [reply]
I thought as much, that's the approach I'm taking currently, just wanted to be sure I wasn't missing an (even) easier solution. Thanks for developing these packages, using them so far has been a pleasure 😃

**Comment by marcemarr** [comment]
Hello, thank you so much for the article! 

Just have an existencial doubt:
Is your function some how related to the content of the tsoutliers package in R?

https://cran.r-project.org/web/packages/tsoutliers/tsoutliers.pdf

Thank you again

**Reply by robjhyndman** [reply]
Not really. My function came first, and then the tsoutliers package was produced. It is a more sophisticated take on the problem.

**Comment by cyrilou242** [comment]
Hello, 
The question from CrossValidated mentionned in the article asks for online detection. 
My understanding of a (simple) online detection setup, is that an anomaly detection is run at each new available point. This is similar to the description in the question "each time a new sample is captured". 

In my tests, I find the `tsoutliers` to perform very poorly in such setup. It seems this method is only appropriate for offline detection.   

Here is a link to a simple test that simulates a repeated period of 1 2 3 4 5 6 7: 
https://gist.github.com/cyrilou242/a979dd8d01fca62b14b1fced36ea6311
After 4 periods -  I consider this is enough for fitting - an anomaly, 10 instead of 2, is inserted at index 30.
The `tsoutliers` is able to detect the anomaly in an offline setup, but in an online setup, when it's the rightmost value, `tsoutliers` does not detect the anomaly and becomes unstable for the following points. It is only able to detect it correctly after 5 more observations are available. 

Intuitively, it's most likely caused by mstl and supmus implementation that have their windows left and right looking. They make assumptions/approximations at the rightmost of the timeseries (where the right window is not available) that are not relevant for online detection. Online detection is most interested in the latest points. 

Am I missing something?
Thanks,
Cyril

**Reply by robjhyndman** [reply]
I think what you want here is an algorithm that fits a model to data over times 1,...,T, and then produces a forecast distribution for time T+1. If the observation at time T+1 is not consistent with the forecast distribution (e.g., the density at the observation is very low), then it is an outlier. This is a different problem from what either the `tsoutliers` package, or the `forecast::tsoutliers()` function are trying to solve.

**Reply by cyrilou242** [reply]
Thanks for clarifying.

> an algorithm that fits a model to data over times 1,...,T, and then produces a forecast distribution for time T+1. 

I did try this approach with an ETS, but the forecasts is sensible to anomalies in 1,...,T, especially with the automatic parameter optimization. Anomalies in 1,..., T also results in too big confidence intervals.
So I'd need to remove the anomalies in the first batch of training data. Kind of a chiken and egg problem.
Any recommendation of a better forecasting method for this use case?

**Reply by cyrilou242** [reply]
Also for the `tsoutliers` problem, the unstability around `[1, frequency/2]` and `[n-frequency/2, n]` could still be an issue I guess. FYI I had correct results by appending data before and after the ts. Eg instead of `tsoutliers(ts(x, frequency=7))`, running `tsoutliers(ts(c(prefix, x, suffix), frequency=7))`. With `prefix` and `suffix` kind of a mirror image of x, respecting the seasonality patterns.

**Reply by robjhyndman** [reply]
If you have anomalies in the training data, you probably need to use a robust model. The robets package is one possibility, but it is currently archived on CRAN. robustarima is another possibility. Or you could simply remove the identified anomalies in the training data by replacing them with missing values. Then use `auto.arima()`.

**Reply by cyrilou242** [reply]
Ok thanks a lot! Will try these approaches.

**Reply by cyrilou242** [reply]
Hey, I incorporated robets ideas in my company ETS implementation and I have good results in real-life use cases.
Thanks a lot for your help!

**Comment by Dast18ac** [comment]
Hi robjhyndman, 

Let me know if I'm missing anything. 

*Any outliers identified in this manner are replaced with linearly interpolated values using the neighbouring observations, and the process is repeated.*

This is problematic for my application as outliers tend to follow eachother. I'm dealing with physical measurements where e.g., a power outage may cause non-zero erroneous for two days in a row. 

Is there a way to adjust the calculation of the replacement value?

Best,
Dan

**Reply by robjhyndman** [reply]
First, if there are consecutive outlying values, `tsclean()` will replace them both by interpolating between the preceding and following non-outlying values.

But if that's not what you want, then `tsoutliers()` identifies the location of outliers and returns a suggested replacement value. So you can easily write your own function to call `tsoutliers()` to find the locations, but then replace the outliers with your own preferred value.

**Reply by Dast18ac** [reply]
Thanks! This helped on my understanding of the function. 

A quick follow up question: I find that the threshold for “far out” values is a bit too loose. Is there a way to adjust this so that observations are labeled as outliers more frequently?

**Reply by robjhyndman** [reply]
Just create your own version of the function and modify the relevant lines

**Comment by ashley2876** [comment]
Hello! Thank you for all of the information you post to this site, it is extremely useful. May I ask, if my outlier values are real and I do not alter them, the impact it will have on the performance of a time series method? My understanding is that methods like STL are robust to outliers, but I'm not sure if this is the case with methods like ETS or ARIMA. Is it better to preserve the real values, or to remove an outlier to try to improve a forecast's performance?

**Reply by robjhyndman** [reply]
STL is robust if you include the `robust=TRUE` argument, but is not otherwise. Neither ETS nor ARIMA in the forecast or fable packages are robust to outliers, but there have been variations proposed that are. Some implementations are available in R, but not on CRAN. See https://github.com/RubenCrevits/robets and https://github.com/spkaluzny/robustarima for example.

If you think outliers are a problem, they can be replaced with missing values, and then the ARIMA function in fable, or the auto.arima function in forecast, will work as expected. Unfortunately, the ETS implementations don't handle missing values.

**Reply by ashley2876** [reply]
Thank you so much!

**Comment by mariospelekanos** [comment]
I am facing the following issue. My time series has a 8 year period of serious increase followed by a similar year period of vast decrease. Afterwards the data points don't have a serious deviation for the next 5 years. By omitting the first ten year data points I noticed a similar cross validation error with all data and with reduced data although different models were used.
If a time series has a large standard deviation would it be a good idea to ignore some data points in order to reduce it?
For example reduce standard deviation by 8% by ignoring first time series observations. Is it safe if half of the data points have to be omitted?

Thank you in advance.

**Reply by robjhyndman** [reply]
I can't offer help about specific problems. Perhaps post a reproducible example to https://forum.posit.co/

**Comment by tyy231027** [comment]
Thanks for this post. Very helpful.

**Comment by theforecastnoob** [comment]
Hello! 
How to think when using outlier detection in hierarchical datasets? 
For example, the top level might have some outliers (according some method), but then it doesn't detect them in the lower levels, or vice versa. (aggregated "lower level- residuals" might become outliers at the top level, or vice versa). 

Is it then more appropriate to do the outlier detection in the same manner as the reconcilation strategy? 
E.g., if we're going to use bottom up reconciliation- we only do outlier detection at the bottom level, and then aggregate the decompositions(trend, seasonal, residual) to obtain the adjusted top level?

Another question, how to tackle when we have intermittent data, any suggestions?
BR, 
J

**Reply by theforecastnoob** [reply]
My last question, I meant - how to do outlier detection with intermittent data*. 
As this can be the case for the lower level in a hierarchical dataset. 
thanks in advance

**Reply by robjhyndman** [reply]
In my view, anomaly detection needs an assumed probability distribution or stochastic model. Then anomalies are points of unusually low density. For time series, this probably means fitting a model to the data, with an assumed distribution for the residuals, then finding points with residuals having very low density. e.g., you could use STL and find points in the tail of the remainder distribution. When you have hierarchical data, with aggregation constraints, it is possible for an anomaly to occur in only some series in the hierarchy under this way of thinking. e.g., large positive remainders in one series may be offset by large negative remainders in a sibling series, and then the aggregate does not contain an anomaly. Or the child series may contribute only a small amount to its parent, so an anomaly in the child does not lead to a detectable anomaly in the parent. So I would be just looking one series at a time and don't worry about the hierarchical constraints.

For intermittent data, again you need a stochastic model. e.g, a Poisson AR model if your time series consists of small non-negative counts. Then you identify anomalies as points of unusually low density.

**Reply by theforecastnoob** [reply]
Thanks!
